storm-core/src/jvm/backtype/storm/blobstore/LocalFsBlobStore.java

```diff
@@ -0,0 +1,306 @@
+    checkForBlobOrDownload(key);
+    validateKey(key);
+    _aclHandler.normalizeSettableBlobMeta(key, meta, who, ADMIN);
+    BlobStoreAclHandler.validateSettableACLs(key, meta.get_acl());
+    SettableBlobMeta orig = getStoredBlobMeta(key);
+    _aclHandler.validateACL(orig.get_acl(), ADMIN, who, key);
+    BlobStoreFileOutputStream mOut = null;
+    try {
+      mOut = new BlobStoreFileOutputStream(fbs.write(META_PREFIX+key, false));
+      mOut.write(Utils.thriftSerialize((TBase) meta));
+      mOut.close();
+      mOut = null;
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    } finally {
+      if (mOut != null) {
+        try {
+          mOut.cancel();
+        } catch (IOException e) {
+          //Ignored
+        }
+      }
+    }
+  }
+
+  @Override
+  public void deleteBlob(String key, Subject who) throws AuthorizationException, KeyNotFoundException {
+    checkForBlobOrDownload(key);
+    validateKey(key);
+    SettableBlobMeta meta = getStoredBlobMeta(key);
+    _aclHandler.validateACL(meta.get_acl(), WRITE, who, key);
+    try {
+      fbs.deleteKey(DATA_PREFIX+key);
+      fbs.deleteKey(META_PREFIX+key);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public InputStreamWithMeta getBlob(String key, Subject who) throws AuthorizationException, KeyNotFoundException {
+    if(!checkForBlobOrDownload(key)) {
+      checkForBlobUpdate(key);
+    }
+    validateKey(key);
+    SettableBlobMeta meta = getStoredBlobMeta(key);
+    _aclHandler.validateACL(meta.get_acl(), READ, who, key);
+    try {
+      return new BlobStoreFileInputStream(fbs.read(DATA_PREFIX+key));
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Iterator<String> listKeys(Subject who) {
+    try {
+      return new KeyTranslationIterator(fbs.listKeys(), DATA_PREFIX);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void shutdown() {
+  }
+
+  @Override
+  public int getBlobReplication(String key, Subject who) throws Exception {
+    validateKey(key);
+    SettableBlobMeta meta = getStoredBlobMeta(key);
+    _aclHandler.validateACL(meta.get_acl(), READ, who, key);
+    CuratorFramework zkClient = Utils.createZKClient(conf);
+    if (zkClient.checkExists().forPath(BLOBSTORE_SUBTREE + key) == null) {
+       zkClient.close();
+       return 0;
+    }
+    int replicationCount = zkClient.getChildren().forPath(BLOBSTORE_SUBTREE + key).size();
+    zkClient.close();
+    return replicationCount;
+  }
+
+  @Override
+  public int updateBlobReplication(String key, int replication, Subject who) throws AuthorizationException, KeyNotFoundException {
+    int replicationCount = 0;
+    validateKey(key);
+    SettableBlobMeta meta = getStoredBlobMeta(key);
+    _aclHandler.validateACL(meta.get_acl(), READ, who, key);
```

Comment:
This should be `WRITE`, not `READ`.  The supervisor should not be allowed to update the blob replication.